### PR TITLE
OADP-3504: imagePullPolicy based on image having or not digest

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

First part of linked JIRA issue. For OADP 1.3, change imagePullPolicy based on container image. If it contains a sha256 or sha 512 digest, use IfNotPresent; otherwise, Always. User can not change that value manually, reconcile loop will just update value.

> **Note:** in production, we already set OADP CSV to IfNotPresent. But it can be changed manually by user.

## How the changes were made

Run
```sh
grep -Inr 'ImagePullPolicy' . --exclude=\*_test.go
grep -Inr 'imagePullPolicy' .
```

## How to test the changes made

Run `make deploy-olm` on this branch and check imagePullPolicy when using different images. Update `RELATED_IMAGE_` environment variables in OADP CSV or use unsupportedOverrides to do that.

Change regex to something wrong to test logs.

### CODE REVIEW

File: pkg/common/common.go
Comment (mateusoliveira43) [inline pkg/common/common.go:223]: reference I found good https://stackoverflow.com/questions/37861791/how-are-docker-image-names-parsed

File: controllers/velero.go
Comment (kaovilai) [reply controllers/velero.go:284]: We can amend this comment by saying it is ok to not always pull for digest images.

This was during dev where sometimes some people forgot to push their changes to the tag and ended up testing older change.

File: pkg/common/common.go
Comment (kaovilai) [inline pkg/common/common.go:224]: Add unit tests?

**Comment by rayfordj** [comment]
no concerns for 1.3 downstream with this PR's changes :sunglasses: :+1:

File: pkg/common/common.go
Comment (shubham-pampattiwar) [inline pkg/common/common.go:225]: To reinforce this logic lets check something like `@sha256` or a regex expression like `@sha256:[a-f0-9]{64}`

Comment (mateusoliveira43) [reply pkg/common/common.go:225]: my fear of being more focused in sha256 here is ignoring another algorithms and digests, or that's ok?

Comment (weshayutin) [reply pkg/common/common.go:225]: @rayfordj would be the authority here, but afaik our images are only shipped w/ @sha256

Comment (rayfordj) [reply pkg/common/common.go:225]: @weshayutin , I believe at this time, it should be relatively safe to be more explicit in the match, as @shubham-pampattiwar suggests, to reinforce the logic.  We -- all of Red Hat using digests -- are currently embedded with sha256 digests for pinned images; in the event image digest does eventually move away from sha256 to something else, we should have very sufficient runway to accommodate.  

@mateusoliveira43 , if it eases your concern, it might be that a regex offers more flexibility for adjustment in the future to match other such algorithms & digests should it be required versus "is `@sha256` in string".  I defer to you.

Comment (rayfordj) [reply pkg/common/common.go:225]: It's times like these I'm reminded of [this](https://xkcd.com/1171/):

Comment (mateusoliveira43) [reply pkg/common/common.go:225]: alright, going to update!

File: controllers/velero.go
Comment (mateusoliveira43) [reply controllers/velero.go:284]: updated comment telling to use new function

File: pkg/common/common.go
Comment (mateusoliveira43) [reply pkg/common/common.go:224]: done

File: pkg/common/common.go
Comment (mateusoliveira43) [reply pkg/common/common.go]: done

**Review by kaovilai — DISMISSED** [review]

File: pkg/common/common.go
Comment (mpryc) [reply pkg/common/common.go:223]: I think better source of reference is here:
   https://github.com/opencontainers/image-spec/blob/main/descriptor.md#digests 
   
   Based on that @sha256 is fine, maybe another validation if the followed set of characters is exactly 64 ?
   
   https://github.com/opencontainers/image-spec/blob/main/descriptor.md#sha-256
   
   Also would be wise to include sha512, because it's part of the spec ?

**Review by mpryc — APPROVED** [review]
I am approving it with one question that does not need to be addressed in this PR (sha512 and additional validation)

**Review by shubham-pampattiwar — DISMISSED** [review]

File: pkg/common/common.go
Comment (mateusoliveira43) [reply pkg/common/common.go:223]: like the reference

will wait feedback about length check to update both, if agreed

**Comment by mateusoliveira43** [comment]
/hold
wait to answer Michal's question (other wise will be 3 more PRs: here, oadp-1.3 and oadp-1.4)

**Comment by kaovilai** [comment]
based on jira, this doesn't seems like something that would port to oadp-1.3?

**Comment by mateusoliveira43** [comment]
this PR is focused more in 1.3 (but also going to 1.4). Next one (same issue), about adding option to choose policy, its only 1.4

**Comment by shubham-pampattiwar** [comment]
@kaovilai next PR will consist of CRD changes to accept user input for imagepullpolicy, this PR is intended as such so that we get a clean cherry pick for oadp-1.3 without the CRD changes

**Comment by kaovilai** [comment]
Fix version in jira needs to change to 1.3 then

**Comment by kaovilai** [comment]
or we need a clone jira linked that fixes 1.3

**Comment by weshayutin** [comment]
We need 5 total pr's 
1. [imagePullPolicy based on image having or not digest](https://github.com/openshift/oadp-operator/pull/1397#top)
2. cherry-pick  ^ to 1.4 - issue https://issues.redhat.com/browse/OADP-3504
3. cherry-pick ^ to 1.3 - issue https://issues.redhat.com/browse/OADP-4171
4. master - add dpa user overide - issue https://issues.redhat.com/browse/OADP-4172
5. cherry-pick ^ to 1.4 - issue: https://issues.redhat.com/browse/OADP-4172

**Comment by rayfordj** [comment]
@mateusoliveira43 , is your hold still relevant?

**Comment by shubham-pampattiwar** [comment]
@mateusoliveira43 If you want to make the PR more robust try using the regex I suggested earlier: https://github.com/openshift/oadp-operator/pull/1397#discussion_r1617663873 
In addition to it for SHA 512 you would need something like `sha@512:[a-f0-9]{128}`

**Comment by weshayutin** [comment]
/hold
holding all 1.4.x patches to wait for ACM updates.

File: pkg/common/common.go
Comment (mateusoliveira43) [reply pkg/common/common.go:223]: added length and sha512 checks

File: pkg/common/common.go
Comment (mateusoliveira43) [inline pkg/common/common.go:226]: should we at least log if an error occurs?

Comment (kaovilai) [reply pkg/common/common.go:226]: Sure.

File: pkg/common/common.go
Comment (shubham-pampattiwar) [inline pkg/common/common.go:225]: lets log the error here or you can ignore the returned error `matched, _ := regexp.Match...` or even better lets use `regex.Compile` first then use `regex.Match`

Comment (mateusoliveira43) [reply pkg/common/common.go:225]: going to take a look in `regex.Compile`

Comment (mateusoliveira43) [reply pkg/common/common.go]: done

had to change more things then I though :cry: 

But tested and looks it is working. Example log from oadp controller manager when changing regex wrongly 
```
2024-06-10T18:14:05Z	ERROR	imagePullPolicy regex failed	{"controller": "dataprotectionapplication", "controllerGroup": "oadp.openshift.io", "controllerKind": "DataProtectionApplication", "DataProtectionApplication": {"name":"velero-sample","namespace":"test-oadp-operator"}, "namespace": "test-oadp-operator", "name": "velero-sample", "reconcileID": "c306a4fd-3d41-48a2-ad38-fadef0e6da2d", "error": "error parsing regexp: missing closing ]: `[a-f0-9{64}`"}
github.com/openshift/oadp-operator/controllers.(*DPAReconciler).customizeNodeAgentDaemonset
	/go/src/github.com/openshift/oadp-operator/controllers/nodeagent.go:335
github.com/openshift/oadp-operator/controllers.(*DPAReconciler).buildNodeAgentDaemonset
	/go/src/github.com/openshift/oadp-operator/controllers/nodeagent.go:228
github.com/openshift/oadp-operator/controllers.(*DPAReconciler).ReconcileNodeAgentDaemonset.func1
	/go/src/github.com/openshift/oadp-operator/controllers/nodeagent.go:145
sigs.k8s.io/controller-runtime/pkg/controller/controllerutil.mutate
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/controller/controllerutil/controllerutil.go:426
sigs.k8s.io/controller-runtime/pkg/controller/controllerutil.CreateOrPatch
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/controller/controllerutil/controllerutil.go:361
github.com/openshift/oadp-operator/controllers.(*DPAReconciler).ReconcileNodeAgentDaemonset
	/go/src/github.com/openshift/oadp-operator/controllers/nodeagent.go:123
github.com/openshift/oadp-operator/controllers.ReconcileBatch
	/go/src/github.com/openshift/oadp-operator/controllers/dpa_controller.go:218
github.com/openshift/oadp-operator/controllers.(*DPAReconciler).Reconcile
	/go/src/github.com/openshift/oadp-operator/controllers/dpa_controller.go:92
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:119
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:316
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227
```

**Comment by weshayutin** [comment]
> based on jira, this doesn't seems like something that would port to oadp-1.3?

someone isn't listening during the scrum :/  Let's GO!

**Review by kaovilai — APPROVED** [review]

**Review by mpryc — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

**Pull request merged**
